libp2p-comit/src/protocol.rs

```diff
@@ -48,3 +50,173 @@ impl OutboundUpgrade<NegotiatedSubstream> for Config {
         future::ok(framed)
     }
 }
+
+/// Represents a prototype for an upgrade to handle the secret hash sharing
+/// protocol.
+///
+/// The protocol works in the following way:
+///
+/// - Dialer (Bob) reads the `SecretHashInfo` from the substream.
+/// - Listener (Alice) writes the `SecretHashInfo` to the substream.
+#[derive(Debug, Clone)]
+pub struct SecretHashProtocolConfig;
+
+/// Data sent to peer in secret hash protocol.
+#[derive(Deserialize, Debug, Serialize)]
+pub struct SecretHashInfo {
+    swap_id: String,     // FIXME: Use SwapId type
+    secret_hash: String, // FIXME: Use secret type from cnd?
+}
+
+/// The substream on which a reply is expected to be sent.
+pub struct ReplySubstream<T> {
+    inner: T,
+}
+
+impl<T> fmt::Debug for ReplySubstream<T> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_tuple("ReplySubstream").finish()
+    }
+}
+
+impl<T> ReplySubstream<T>
+where
+    T: AsyncWrite + Unpin,
+{
+    /// Sends back the requested information on the substream.
+    ///
+    /// Consumes the substream, returning a `ReplyFuture` that resolves
+    /// when the reply has been sent on the underlying connection.
+    pub fn send(mut self, info: SecretHashInfo) -> impl Future<Output = Result<(), io::Error>> {
+        debug!("Sending secret hash info to peer");
+
+        let message = serde_json::to_vec(&info).unwrap();
+
+        async move { upgrade::write_one(&mut self.inner, &message).await }
+    }
+}
+
+impl UpgradeInfo for SecretHashProtocolConfig {
+    type Info = &'static [u8];
+    type InfoIter = iter::Once<Self::Info>;
+
+    fn protocol_info(&self) -> Self::InfoIter {
+        iter::once(b"/comit/swap/secret_hash/1.0.0") // FIXME: add a helper
+                                                     // function to generate
+                                                     // this.
+    }
+}
+
+impl<C> InboundUpgrade<C> for SecretHashProtocolConfig
+where
+    C: AsyncRead + AsyncWrite + Unpin,
+{
+    type Output = ReplySubstream<C>;
```

Comment:
I've seen this design in the libp2p codebase and always wondered what the motivation is.

Do you think we can figure this out?
More concretely, what is the reason that `InboundUpgrade` cannot mirror `OutboundUpgrade` and just write to the socket directly instead of reading?

The consequence would be that `SecretHashProtocolConfig` would have to have the `SecretHashInfo` as a field so it can send it. That is obviously useless for the `OutboundUpgrade` trait. But those two don't necessarily have to be implemented on the same struct. There could be one protocol config for inbound and one for outbound.

We should experiment with this idea to understand the need for this particular design because at the moment, I don't see a specific one.

Reply:
> I've seen this design in the libp2p codebase and always wondered what the motivation is.
>
> Do you think we can figure this out?

I've got a feeling it will work out nicer if we follow libp2p's example but use an immediately ready future for the outbound upgrade that resolves to a type that we can call send on to send the data (instead of putting it in the config object).  I've not got a fully formed argument for this so I'll explore the idea further then either comment more or post a follow up PR.

Reply:
I would like us to figure out why we are doing this indirection before going for it. Opening a substream is going to require us to create a `ProtocolsHandlerEvent::OutboundSubstreamRequest` like here: https://github.com/comit-network/comit-rs/blob/76c7e4421fba55c903256bf69cb4be6e8edd2b43/libp2p-comit/src/handler.rs#L189-L192

Here we instantiate the config and pass some info along. The info is passed back to us in `inject_fully_negotiated_outbound`:

https://github.com/comit-network/comit-rs/blob/76c7e4421fba55c903256bf69cb4be6e8edd2b43/libp2p-comit/src/handler.rs#L141-L157

I don't see the point of calling `send` on the first argument with the `info` that is passed along. The upside I see is that we can have the same `Config` implementation for both, upgrading inbound and outbound.

I feel like it should be symmetric: Either we only upgrade and return a substream in both cases or we send/receive in both cases.

I would be keen to do the latter because it means we have libp2p-core driving the future to completion instead of us having to store it inside the `ProtocolsHandler` implementation and manually poll it.

Reply:
Ok, I missed this comment until now and have not done as suggested while asking for 50% review.  Will reconsider now I've read this.